File: src/leetCodeProblems/Sorting/MajorityElementII229.java
package leetCodeProblems.Sorting;

/**
 * LeetCode - https://leetcode.com/problems/majority-element-ii/
 */
import java.util.*;

public class MajorityElementII229 {
	
	public List<Integer> majorityElement(int[] nums) {
        
		Arrays.sort(nums);        
        
        List<Integer> outputList = new ArrayList<Integer>();
        
        int localCurrentCount = 0;
        int localCurrentElement = 0;
            
        for(int i=0; i<nums.length; i++) {
            
            if (nums[i] != localCurrentElement) {
                localCurrentElement = nums[i];
                localCurrentCount = 0;
            }
            
            localCurrentCount++;
            
            if (localCurrentCount > (nums.length/3) && !outputList.contains(nums[i])) {
                outputList.add(nums[i]);
            }
        }
        
        return outputList;
        
    }
    
    public static void main(String[] args) {
        
    	//int[] inputArray = {1};
    	
        //int[] inputArray = {3, 2, 3};
        
        int[] inputArray = {2,2,1,1,1,2,2};
        
        MajorityElementII229 obj = new MajorityElementII229();
        
        System.out.println(obj.majorityElement(inputArray));
    }

}

File: src/leetCodeProblems/Sorting/MajorityElementII229.go
package main

import (
	"sort"
)

/*
- LeetCode - https://leetcode.com/problems/majority-element-ii/
- Time - O(nlogn)
- Space - O(1)
*/

func majorityElement(nums []int) []int {
	out := []int{}

	sort.Ints(nums)

	current, count := 0, 0

	for _, v := range nums {
		if v != current {

			if count > len(nums)/3 {
				out = append(out, current)
			}

			current, count = v, 0
		}

		count++
	}

	if count > len(nums)/3 {
		out = append(out, current)
	}
	return out
}

// func main() {
// 	log.Println(majorityElement([]int{3, 2, 3}))
// 	log.Println(majorityElement([]int{1}))
// 	log.Println(majorityElement([]int{1, 2}))
// }
